### PR TITLE
Polars: boolean functions

### PR DESCRIPTION
**Pull request by Shoeboxam**
Adds support for the following transformations:
* is_null
* is_not_null
* is_finite
* is_infinite
* is_nan
* is_not_nan
* not

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1586?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1586** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1586?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1584** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1584?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1498** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1534](https://github.com/opendp/opendp/pull/1534) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1534?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1479** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1479?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1496](https://github.com/opendp/opendp/pull/1496) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#908** <a href="https://app.graphite.dev/github/pr/opendp/opendp/908?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#913](https://github.com/opendp/opendp/pull/913) <a href="https://app.graphite.dev/github/pr/opendp/opendp/913?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#890** <a href="https://app.graphite.dev/github/pr/opendp/opendp/890?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#856](https://github.com/opendp/opendp/pull/856) <a href="https://app.graphite.dev/github/pr/opendp/opendp/856?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#829** <a href="https://app.graphite.dev/github/pr/opendp/opendp/829?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#854](https://github.com/opendp/opendp/pull/854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#888](https://github.com/opendp/opendp/pull/888) <a href="https://app.graphite.dev/github/pr/opendp/opendp/888?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1519** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1519?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1506](https://github.com/opendp/opendp/pull/1506) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1506?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1566** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1566?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/transformations/make_stable_expr/expr_boolean_function/mod.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        ..
+    } = expr
+    else {
+        return fallible!(MakeTransformation, "expected boolean function expression");
+    };
+
+    let Ok([input]) = <&[_; 1]>::try_from(input.as_slice()) else {
+        return fallible!(
+            MakeTransformation,
+            "{} must have one argument",
```
[the remark this answers is not in the document]
Comment (mccalluc) [reply rust/src/transformations/make_stable_expr/expr_boolean_function/mod.rs:44]: What would happen if someone tried to use `is_between`? I'm guessing that that counts as a boolean function, but I also would guess it has two arguments, so the message here would be wrong.

Maybe we want a match for whether we support the function before checking arguments?

I could be going in the wrong direction here.

File: rust/src/transformations/make_stable_expr/expr_boolean_function/mod.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    use BooleanFunction::*;
+
+    match bool_function {
+        IsNull | IsNotNull => active_series.nullable = false,
+        IsFinite | IsInfinite | IsNan | IsNotNan | Not => (),
+        _ => {
+            return fallible!(
+                MakeTransformation,
+                "{:?} is not a recognized row-by-row boolean function",
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_boolean_function/mod.rs:65]: In another PR, I think the messaging emphasized that a failed match just wasn't something we'd implemented yet. This makes it sound more like a user error.

File: rust/src/transformations/make_stable_expr/expr_boolean_function/mod.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                bool_function
+            )
+        }
+    }
+
+    if bool_function == Not && active_series.field.dtype != DataType::Boolean {
+        active_series.drop_bounds()?;
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_boolean_function/mod.rs:72]: Confused here: If they try to use `not`, but the data isn't `boolean`, should that be an error? Or is that not what's going on here?

File: rust/src/transformations/make_stable_expr/expr_boolean_function/test.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        SeriesDomain::new("i64_null", OptionDomain::new(AtomDomain::<i64>::default())),
+    ])?;
+
+    let lf = df!(
+        "f64" => [0., f64::NAN, f64::NAN, f64::INFINITY],
+        "i64" => [0, 1, 2, 3],
+        "f64_null" => [Some(0.), Some(f64::NAN), None, Some(f64::INFINITY)],
+        "i64_null" => [Some(0), None, None, Some(3)],
+    )?
+    .lazy();
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_boolean_function/test.rs:20]: This is a matter of taste, but even if it's more repetitive, there's something to be said for having the input for a unit test in that same unit test, and being able to glace at a small number of contiguous lines and confirm that the output is what we'd expect.

File: rust/src/transformations/make_stable_expr/expr_boolean_function/test.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+    Ok((lf_domain, lf))
+}
+
+macro_rules! is_nullable {
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_boolean_function/test.rs:25]: In another PR there's a comment about nullable vs. nanable (for lack of a better name). I'm not sure if this is something that would change.

**Review by mccalluc — COMMENTED** [review]
As more of the polars feature space is filled in, I worry a little about keeping track of what isn't implemented, both to avoid user surprise, and to make sure we haven't forgotten anything. If we've got most of an enumeration covered, maybe include matches for the full set rather than a fall through case, some the coverage is more clear?

File: rust/src/transformations/make_stable_expr/expr_boolean_function/test.rs (hunk 4 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_boolean_function/test.rs:20]: That would be an awful lot of duplication...

**Comment by Shoeboxam** [comment]
I think it is unlikely that users will find a whitelist specifically of boolean functions useful. A global whitelist would be useful, though. The fall-through is also helpful to avoid needing to enable all features.

**Comment by mccalluc** [comment]
From the description:
> This is a building block for supporting nan imputation.

but it is being merged into [1577-impute-nan](https://github.com/opendp/opendp/tree/1577-impute-nan) ... which makes it seems like it's building on top of nan imputation, instead of being the first step.

**Comment by Shoeboxam** [comment]
> From the description:
> 
> > This is a building block for supporting nan imputation.
> 
> but it is being merged into [1577-impute-nan](https://github.com/opendp/opendp/tree/1577-impute-nan?rgh-link-date=2024-06-13T21%3A16%3A20Z) ... which makes it seems like it's building on top of nan imputation, instead of being the first step.

Ah, I implemented this first and ultimately decided I couldn't use these to build fill_nan, because you need to parse the whole ternary at once. If you parse the whole ternary for fill_nan at once, you can see that it fills all nans with non-nan fill values. However, if you just parse the ternary, and then recursively the condition, then you don't have the info you need to set the output domain's descriptor to non-nan. 

I finished this, and then opened separate PRs and moved them down the stack because they were more relevant for FSO.

**Review by mccalluc — APPROVED** [review]
Spot check looks good. Thanks!

**Pull request merged**